README
Installation de CTM::* :

- perl Makefile.PL
- make
- make install
- make clean

Sources disponibles sur :

- http://search.cpan.org/dist/CTM
- http://github.com/WeebleDev/CTM

Pour toutes autres informations, voir la POD du module ReadEM.pm

CHANGES
* 0.09 et moins (STABLE) :
    * < 23/03/2014 - tests.
* 0.10 (STABLE) :
    * 24/03/2014 - premiere version stable.
* 0.11 (STABLE) :
    * 25/03/2014 - support de Control-M EM v6.
    * 25/03/2014 - mise en place du module _BIMServices dont le constructeur est une methode de Control-M EM::BIM.
    * 31/03/2014 - methodes du module _BIMServices misent en place.
    * 07/04/2014 - modifications des methodes liees au module Control-M EM::BIM et aux services : la recuperation des services se fait desormais dynamiquement via l'active net.
    * 17/04/2014 - refonte de la doc POD du module et d'une partie de la gestion des erreurs via Carp.
    * 23/04/2014 - gestion des jobs supprimes du plan via la cle 'delete_flag' (booleen) et avec le parametre 'handleDeletedJobs' de la methode getCurrentServices() et de ses derives.
* 0.12 (STABLE) :
    * 24/04/2014 - renommage du module _BIMServices en _WorkOnServices.
    * 24/04/2014 - refonte de l'espace de nom du module (Control-M EM::BIM -> CTM::EM::BIM::ServicesAPI) et de son nom (BIM.pm -> ServicesAPI.pm).
* 0.13 (STABLE) :
    * 07/05/2014 - ajout de la methode getSOAPEnvelope() au module _WorkOnServices.
    * 07/05/2014 - securisation des objets via Hash::Util et ajout d'un mutateur (setPublicProperty()) pour les proprietes publiques des objets.
    * 09/05/2014 - ajout de la methode getProperty().
* 0.14 (STABLE) :
    * 09/05/2014 - module CTM::EM::BIM::ServicesAPI::Common fonctionnel. Contient les fonctions/methodes _setObjProperty(), _myErrorMessage(), getProperty(), setPublicProperty() et clearError().
    * 09/05/2014 - heritage de certaines fonctions pour les modules CTM::EM::BIM::ServicesAPI et _WorkOnServices depuis CTM::EM::BIM::ServicesAPI::Common.
    * 20/05/2014 - ajout du parametre "verbose" a la methode new().
    * 20/05/2014 - ajout de la methode refresh() au module _WorkOnServices.
* 0.15 (STABLE) :
    * 22/05/2014 - refonte de l'espace de nom pour les modules CTM::* (CTM::Base, CTM::ReadEM, CTM::ReadEM::_workOnBIMServices).
    * 22/05/2014 - certaines fonctions privees deviennent accessibles (_myOsIsUnix(), ...).
* 0.161 (STABLE) :
    * 26/05/2014 - ajout du parametre 'forStatus' a la methode getCurrentServices().
    * 26/05/2014 - POD revue pour l'upload des modules sur le CPAN.
* 0.161 (STABLE) :
    * 26/05/2014 - correction de quelques bugs relatifs a la methode $_setObjProperty->().
* 0.162 (STABLE) :
    * 26/05/2014 - heritage d'AUTOLOAD depuis CTM::Base.
* 0.17 (STABLE) :
    * 30/05/2014 - ajout des modules CTM::ReadEM::_workOnAlarms, CTM::ReadEM::_workOnExceptionAlerts et CTM::Base::subClass + impacts majeurs sur les modules CTM::ReadEM, CTM::Base et  CTM::ReadEM::_workOnBIMServices.
* 0.171 (STABLE)
    * 31/05/2014 - correction de quelques erreurs dans la POD de CTM::ReadEM et ajout du tag ':all' (%EXPORT_TAG) a ce meme module.
* 0.172 (STABLE)
    * 01/06/2014 - correction de la methode protegee _refresh().
    * 01/06/2014 - correction de plusieurs erreurs pour les methodes getAlarms() et getExceptionAlerts() (parametre 'severity').
    * 01/06/2014 - limitation de l'utilisation de regex (gain de performance tres faible).
* 0.173 (STABLE)
    * 02/06/2014 - apport des parametres "limit" et "timeSort" aux methodes getAlarms() et getExceptionAlerts() ainsi qu'a leurs derives.
    * 02/06/2014 - amelioration de la verification de certains parametres pour les methodes new() et get*.
* 0.174 (STABLE)
    * 02/06/2014 - correction de la methode protegee _refresh().
* 0.175 (STABLE)
    * 02/06/2014 - appelle de la methode unshiftError() lors de l'appelle des methodes getAlerts(), getProblematicsJobs() et getSOAPEnvelope().
* 0.176 (STABLE)
    * 23/06/2014 - ajout des methodes destinees a la gestion des alarmes/alertes dans les modules CTM::ReadEM::WorkOnAlarms et CTM::ReadEM::WorkOnExceptionAlerts.
    * 23/06/2014 - correction de quelques erreurs.
    * 23/06/2014 - mis a jour de la POD.
* 0.177 (STABLE)
    * 30/06/2014 - correction des requetes SQL DELETE.
* 0.1771 (STABLE)
    * 12/07/2014 - utilisation Perl:OSType et de Try::Tiny.
    * 18/07/2014 - correction typo sur la methode CTM::ReadEM::WorkOnExceptionAlerts::refresh() et sur l'utilisation de substr() dans CTM::Base.
    * 18/07/2014 - ajout du test 01init.t et d'un script d'exemple (get_bim_services.pl).
    * 18/07/2014 - prise en compte du champ 'DAYTIME' avec ses variations (+/-) dans la fonction privee $_calculStartEndDayTimeInPosixTimestamp->() (fonctionnalite experimentale).
    * 20/08/2014 - utilisation des modules List::MoreUtils et Date::Calc.
* 0.18 (DEV)
    * 10/09/2014 - ajout de methodes de 'filtrage' (modification de l'objet) (keepItemsWithAnd() et keepItemsWithOr()) aux modules existants.
    * 12/09/2014 - ajout d'une methode de clonage (clone() a travers $_clone() qui requiert le module Storable) pour les sous-modules de CTM::ReadEM.
    * 13/09/2014 - modification des methodes countItems(), getItems(), getSOAPEnvelope() (modifs + -> generateSOAPEnvelope()), _getFromRequest(), _setSerials() et _refresh() (-> resetAndRefresh()).
    * 24/09/2014 - mise a jour du Makefile.PL.
    * 03/10/2014 - correction de typos, ajout de placeholders sur la quasi-totalite des requetes SQL des modules CTM et ajout de la methode publique getErrors().
    * 24/10/2014 - Renommage de la methode getCurrentServices() en getCurrentBIMServices(), de la methode workOnCurrentServices() en workOnCurrentBIMServices() et du module CTM::ReadEM::WorkOnBIMServices en CTM::ReadEM::WorkOnCurrentBIMServices.
    * 24/10/2014 - ajout des fonctions _myVerboseMessage() et _myUsageMessage(), des methodes protegees _tagAtWork() et _tagAtRest() et utilisation du module Sub::Name. Mise en place de constantes afin de regrouper plusieurs informations a propos d'informations internes aux modules CTM::*.
    * 24/10/2014 - correction de nombreux typos.

lib/CTM/Base.pm
#------------------------------------------------------------------------------------------------------
# OBJET : "Classe abstraite" des modules de CTM
# APPLICATION : Control-M
# AUTEUR : Yoann Le Garff
# DATE DE CREATION : 09/05/2014
#------------------------------------------------------------------------------------------------------
# USAGE / AIDE
#   perldoc CTM::Base
#------------------------------------------------------------------------------------------------------

#-> BEGIN

#----> ** initialisation **

package CTM::Base;

use 5.6.1;

use strict;
use warnings;

use constant {
    _CLASS_INFOS => {
        common => {
            rootClass => {
                name => 'CTM::ReadEM',
                propertyName => '_CTM::ReadEM'
            },
            baseClass => {
                name => 'CTM::Base'
            },
            baseSubClass => {
                name => 'CTM::Base::SubClass'
            },
            objectProperties => {
                verbose => {
                    name => 'verbose'
                },
                working => {
                    name => '_working'
                },
                errors => {
                    name => '_errors'
                },
                DBI => {
                    name => '_DBI'
                },
                sessionIsConnected => {
                    name => '_sessionIsConnected'
                },
                parameters => {
                    name => '_params'
                },
                subClassDatas => {
                    name => '_datas'
                }
            }
        },
        rootClass => {
            classProperties => {
                nbSessionsInstanced => {
                    name => 'nbSessionsInstanced'
                },
                nbSessionsConnected => {
                    name => 'nbSessionsConnected'
                }
            }
        },
        BIMCurrentBIMServices => {
            moduleLastName => 'WorkOnCurrentBIMServices',
            baseMethod => 'getCurrentBIMServices',
            workMethod => 'workOnCurrentBIMServices'
        },
        GASAlarms => {
            moduleLastName => 'WorkOnAlarms',
            baseMethod => 'getAlarms',
            workMethod => 'workOnAlarms'
        },
        GASExceptionAlerts => {
            moduleLastName => 'WorkOnExceptionAlerts',
            baseMethod => 'getExceptionAlerts',
            workMethod => 'workOnExceptionAlerts'
        },
        CMComponents => {
            moduleLastName => 'WorkOnComponents',
            baseMethod => 'getComponents',
            workMethod => 'workOnComponents'
        }
    }
};

use Carp qw/
    carp
    croak
/;
use Hash::Util qw/
    lock_hash
    unlock_hash
    lock_value
    unlock_value
/;

#----> ** variables de classe **

our $VERSION = 0.18;
our $AUTOLOAD;

#----> ** fonctions privees (mais accessibles a l'utilisateur pour celles qui ne sont pas des references) **

sub _myErrorMessage($$) {
    my ($subroutine, $message) = @_;
    return "'" . $subroutine . "()' : " . $message;
}

sub _myUsageMessage($$) {
    my ($namespace, $properties) = @_;
    return 'USAGE : ' . (split(/::/, $namespace))[-1] . '(' . $properties . ').';
}

#----> ** methodes protegees **

sub _invokeVerbose {
    my ($self, $subroutine, $message) = @_;
    if (caller->isa(__PACKAGE__)) {
        printf STDERR "VERBOSE - '%s()' : %s", $subroutine, $message if ($self->{_CLASS_INFOS->{common}->{objectProperties}->{verbose}->{name}});
    } else {
        carp(_myErrorMessage((caller 0)[3], "tentative d'utilisation d'une methode protegee."));
    }
    return 0;
}

#-> accesseurs/mutateurs

sub _setObjProperty {
    my ($self, $property, $value) = @_;
    if (caller->isa(__PACKAGE__)) {
        my $action = exists $self->{$property};
        $action ? unlock_value(%{$self}, $property) : unlock_hash(%{$self});
        $self->{$property} = $value;
        $action ? lock_value(%{$self}, $property) : lock_hash(%{$self});
        return 1;
    } else {
        carp(_myErrorMessage((caller 0)[3], "tentative d'utilisation d'une methode protegee."));
    }
    return 0;
}

sub _isWorking {
    my $self = shift;
    if (caller->isa(__PACKAGE__)) {
        return $self->{_CLASS_INFOS->{common}->{objectProperties}->{working}->{name}};
    } else {
        carp(_myErrorMessage((caller 0)[3], "tentative d'utilisation d'une methode protegee."));
    }
    return 0;
}

sub _tagAtWork {
    my $self = shift;
    if (caller->isa(__PACKAGE__)) {
        return $self->_setObjProperty(_CLASS_INFOS->{common}->{objectProperties}->{working}->{name}, 1);
    } else {
        carp(_myErrorMessage((caller 0)[3], "tentative d'utilisation d'une methode protegee."));
    }
    return 0;
}

sub _tagAtRest {
    my $self = shift;
    if (caller->isa(__PACKAGE__)) {
        return $self->_setObjProperty(_CLASS_INFOS->{common}->{objectProperties}->{working}->{name}, 0);
    } else {
        carp(_myErrorMessage((caller 0)[3], "tentative d'utilisation d'une methode protegee."));
    }
    return 0;
}

sub _addError {
    my ($self, $value) = @_;
    if (caller->isa(__PACKAGE__)) {
        unlock_value(%{$self}, _CLASS_INFOS->{common}->{objectProperties}->{errors}->{name});
        unshift @{$self->getErrors()}, $value;
        lock_value(%{$self}, _CLASS_INFOS->{common}->{objectProperties}->{errors}->{name});
        return 1;
    } else {
        carp(_myErrorMessage((caller 0)[3], "tentative d'utilisation d'une methode protegee."));
    }
    return 0;
}

#----> ** methodes publiques **

#-> accesseurs/mutateurs

sub getProperty {
    my ($self, $property) = @_;
    my $subName = (caller 0)[3];
    croak(_myErrorMessage($subName, _myUsageMessage('$obj->' . $subName, '$definedPropertyName'))) unless (defined $property);
    $self->unshiftError();
    return $self->{$property} if (exists $self->{$property});
    carp(_myErrorMessage($subName, "propriete ('" . $property . "') inexistante."));
    return 0;
}

sub setPublicProperty {
    my ($self, $property, $value) = @_;
    my $subName = (caller 0)[3];
    croak(_myErrorMessage($subName, _myUsageMessage('$obj->' . $subName, '$definedPropertyName, $definedValue'))) unless (defined $property);
    $self->unshiftError();
    unless (exists $self->{$property}) {
        carp(_myErrorMessage($subName, "tentative de creation d'une propriete ('" . $property . "')."));
    } elsif (substr($property, 0, 1) eq '_') {
        carp(_myErrorMessage($subName, "tentative de modication d'une propriete ('" . $property . "') protegee ou privee."));
    } else {
        return $self->_setObjProperty($property, $value);
    }
    return 0;
}

sub getErrors {
    return shift->{_CLASS_INFOS->{common}->{objectProperties}->{errors}->{name}};
}

sub getError {
    my ($self, $arrayItem) = @_;
    return $self->getErrors()->[(defined $arrayItem && $arrayItem =~ /^[\+\-]?\d+$/) ? $arrayItem : 0];
}

sub countErrors {
    my $self = shift;
    return scalar @{$self->getErrors()};
}

sub unshiftError {
    return shift->_addError(undef);
}

sub clearErrors {
    my $self = shift;
    unlock_value(%{$self}, _CLASS_INFOS->{common}->{objectProperties}->{errors}->{name});
    delete @{$self->getErrors()}[0..@{$self->getErrors()}];
    lock_value(%{$self}, _CLASS_INFOS->{common}->{objectProperties}->{errors}->{name});
    return 1;
}

#-> Perl BuiltIn

sub AUTOLOAD {
    my $self = shift;
    if ($AUTOLOAD) {
        no strict qw/refs/;
        (my $called = $AUTOLOAD) =~ s/.*:://;
        croak("'" . $AUTOLOAD . "()' est introuvable.") unless (exists $self->{$called});
        return $self->{$called};
    }
    return undef;
}

1;

#-> END

__END__

=pod

=head1 NOM

CTM::Base

=head1 SYNOPSIS

"Classe abstraite" des modules de C<CTM>.
Pour plus de details, voir la documention POD de C<CTM::ReadEM>.

=head1 DEPENDANCES DIRECTES

C<Carp>

C<Hash::Util>

=head1 NOTES

Ce module est dedie aux modules C<CTM>.

=head1 LIENS

- Depot GitHub : http://github.com/le-garff-yoann/CTM

=head1 AUTEUR

Le Garff Yoann <pe.weeble@yahoo.fr>

=head1 LICENCE

Voir licence Perl.

=cut
